generator/src/main/java/com/scottlogic/deg/generator/generation/StandardFieldValueSourceEvaluator.java

```diff
@@ -105,31 +85,44 @@
                 getBlacklist(fieldSpec)));
         }
 
+        if (mayBeNull(fieldSpec)){
+            validSources.add(nullOnlySource);
+        }
+
         return validSources;
     }
 
-    private boolean determineNullabilityAndDecideWhetherToHalt(
-        Set<FieldValueSource> fieldValueSources,
-        FieldSpec fieldSpec) {
+    private List<FieldValueSource> getWhitelistSources(FieldSpec fieldSpec) {
+        MustContainRestriction mustContainRestriction = fieldSpec.getMustContainRestriction();
+        Stream<Object> whitelist = fieldSpec.getSetRestrictions().getWhitelist().stream();
 
-        FieldValueSource nullOnlySource = new CannedValuesFieldValueSource(Collections.singletonList(null));
+        // If we have values that must be included we need to check that those values are included in the whitelist
+        if (mustContainRestriction != null) {
+            whitelist = Stream.concat(whitelist,
+            getNotNullSetRestrictionFilterOnMustContainRestriction(mustContainRestriction)
+                .flatMap(o -> o.getSetRestrictions().getWhitelist().stream()));
+        }
 
-        if (fieldSpec.getNullRestrictions() != null) {
-            if (fieldSpec.getNullRestrictions().nullness == Nullness.MUST_BE_NULL) {
-                // if *always* null, add a null-only source and signal that no other sources are needed
-                fieldValueSources.add(nullOnlySource);
-                return true;
-            } else if (fieldSpec.getNullRestrictions().nullness == Nullness.MUST_NOT_BE_NULL) {
-                // if *never* null, add nothing and signal that source generation should continue
-                return false;
-            }
+        if (mayBeNull(fieldSpec)) {
+            return Arrays.asList(
+                new CannedValuesFieldValueSource(whitelist.collect(Collectors.toList())),
+                nullOnlySource);
         }
 
-        // if none of the above, the field is nullable
-        fieldValueSources.add(nullOnlySource);
-        return false;
+        return Arrays.asList(new CannedValuesFieldValueSource(whitelist.collect(Collectors.toList())));
     }
 
+    private boolean mayBeNull(FieldSpec fieldSpec) {
+        return fieldSpec.getNullRestrictions() == null;
+    }
+
+    private boolean mustBeNull(FieldSpec fieldSpec) {
+        return fieldSpec.getNullRestrictions() != null
+            && fieldSpec.getNullRestrictions().nullness == Nullness.MUST_BE_NULL;
+    }
+
+    private static final CannedValuesFieldValueSource nullOnlySource = new CannedValuesFieldValueSource(Collections.singletonList(null));
```

Comment:
Formatting comment, you'd normally expect fields, constants, etc. to be at the top of the class, not interweaved with other methods.
Can this be moved to the top please.